release/authorization/README.md

```diff
@@ -0,0 +1,370 @@
+# gNMI Authorization Protocol
+
+## Objective
+
+This proto definition and the reference code(to be delivered seperately) serve
+to describe an authorization framework for controlling which OpenConfig paths of
+a network device specific users can access using a gNMI micro-serivice. The
+authorization policy is initially intended to be deployed to a device, with the
+ability to define:
+
+*   Policy rules - each rule defines a single authorization policy.
+*   Groups of users - as a method to logically group users in the administrative
+    domain, for instance: operators or administrators.
+*   Users which are individual users as part of either rules or in groups.
+    Authentication information is not included in this Authorization
+    configuration.
+
+Policy rules are matched based on the best match for the authorization request,
+not the first match against a policy rule. Best match enables a configuration
+which permits a user or group access to particular OpenConfig paths while
+denying subordinate portions of the permitted paths, or the converse.
+
+Best, or most specific, match is that which has the longest match to the
+requested path and prefers a specific user over a group in the matching
+policy.
+
+Match rules permit a match against:
+
+*   User or Group (not both)
+*   an OpenConfig path
+
+An implicit deny is assumed, if there is no matching rule in the policy. Logging
+may be specified on a per-policy-rule basis as well as a default for the whole
+authorization policy.
+
+[OpenConfig paths](https://github.com/openconfig/reference/blob/master/rpc/gnmi/gnmi-specification.md#222-paths)
+are heirarchical, rooted at a defined "origin". OpenConfig may contain paths
+such as:
+
+```proto
+    /a/b/c/d
+```
+
+Paths may also have attributes associated with the path elements such as:
+
+```proto
+    /a/b[key=foo]/c/d
+```
+
+Attributes may be wildcarded in the policy, such as:
+
+```proto
+    /a/b[key=*]/c/d
+```
+
+Wildcards are only valid in policy rules when applied to attributes.
+Permitted use:
+
+```proto
+    /a/b[key=*]/c/d
+```
+
+Not permitted use:
+```proto
+    /a/b[key=foo]/*/d
+```
+
+## Bootstrap / Install Options
+
+System bootstrap, or install, operations may include an authorization policy
+delivered during bootstrap operations. It is suggested that the bootstrap
+process include the complete authorization policy so all production tools
+and services have immediate authorized access to finish installation and
+move devices into production in a timely manner.
+
+Using the Secure Zero Touch Provisioning (sZTP - RFC8572) process for
+bootstrap/installation provides a method to accomplish this delivery, and
+the delivery of all other bootstrap artifacts in a secure manner.
+
+## Conflict Resolution
+
+Policy evalyation should end with a single best match policy for the provided
```

Comment:
There can be certain ambiguous cases in the current proto. And also, we cannot treat multiple "best match" is due to invalid config. Here are some of the ambiguous cases:

1. What happy if two policy rules conflict with each other (same path, same user, same level but different action)? Should we honor the deny rule, the first rule, the last rule?
Example:
Rule 1: /a/b/c -> allow
Rule 2: /a/b/c -> deny

2. Each policy rule has a level (read_only or read_write), should we match on the exact level? (The examples given in this doc do not have level.)
Example 1:
Rule 1: /a/b/c -> read_only allow
Rule 2: /a/b/c/d -> read_write deny
If we have a read request on /a/b/c/d, should we match on rule 1 or rule 2? The expected behavior is that the user should be able to read /a/b/c/d.
Example 2:
Rule 1: /a/b/c -> read_only deny
Rule 2: /a/b/c/d -> read_write allow
If we have a read request on /a/b/c/d, should we match on rule 1 or rule 2? The expected behavior is that the user should be able to read /a/b/c/d.
Example 3:
Rule 1: /a/b/c -> read_only deny
Rule 2: /a/b/c -> read_write allow
Is the user allow to read /a/b/c? How about write?

3. From #2, what happen if a user matches both a user rule and a group rule?
Example 1:
Rule 1: /a/b/c -> user read_write deny
Rule 2: /a/b/c -> group read_only allow
If a user matches both rules, do they have read access on /a/b/c?

Example 2:
Rule 1: /a/b/c -> user read_write deny
Rule 2: /a/b/c -> group read_write allow
If a user matches both rules, do they have read access on /a/b/c? The intention of the rule is that the group has write access except one user.

4. We also have wildcard. What is "best match" mean?
Example 1:
Rule 1: /a/b[name=mgmt_intf]/c -> deny
Rule 2: /a/b[name=*]/c/d -> allow
If we have a request of /a/b[name=mgmt_intf]/c/d, it matches both rules. Rule 2 has longer prefix. However, the intention of the rule is that we deny access on the management interface (mgmt_intf).

Example 2:
Rule 1: /a/b[k1=*]/c/d[k2=v2]/e -> deny
Rule 2: /a/b[k1=v1]/c/d[k2=*]/e -> allow
If we have a request of /a/b[k1=v1]/c/d[k2=v2]/e, it matches both rules.

Example 3:
Rule 1: /a/b[k1=*][k2=v2]/c/d -> deny
Rule 2: /a/b[k1=v1][k2=*]/c/d -> allow
Multiple keys in the same path element is rare, but it is possible in the proto. A request of /a/b[k1=v1][k2=v2]/c/d will match both rules.

Reply:
> There can be certain ambiguous cases in the current proto. And also, we cannot treat multiple "best match" is due to invalid config. Here are some of the ambiguous cases:
> 
> 1. What happy if two policy rules conflict with each other (same path, same user, same level but different action)? Should we honor the deny rule, the first rule, the last rule?
>    Example:
>    Rule 1: /a/b/c -> allow
>    Rule 2: /a/b/c -> deny

For this case, I'd prefer that we error during ingestion of the config, actually.
This should force the operator to review/test their configuration prior to sending it to the device(s).

The other option to consider is to just follow a relatively standard:
  "Last write wins"

I think this will lead to ambiguity and troubleshooting problems, so I'm in favor of the first which is validation/testing prior to delivery to the device.

> 2. Each policy rule has a level (read_only or read_write), should we match on the exact level? (The examples given in this doc do not have level.)

I'll add some more detail to the examples. (I may steal your examples in fact!)

>    Example 1:
>    Rule 1: /a/b/c -> read_only allow
>    Rule 2: /a/b/c/d -> read_write deny
>    If we have a read request on /a/b/c/d, should we match on rule 1 or rule 2? 
>    The expected behavior is that the user should be able to read /a/b/c/d.

My intent is that the best match here is 'read_write for the longest path', so in your example the
2nd rule should match and cause a DENY to be returned. I bet we need to be able to separately
define READ and WRITE access, and not conjoin RW, I updated the proto definition to reflect that
decision/change.

>    Example 2:
>    Rule 1: /a/b/c -> read_only deny
>    Rule 2: /a/b/c/d -> read_write allow
>    If we have a read request on /a/b/c/d, should we match on rule 1 or rule 2?
>    The expected behavior is that the user should be able to read /a/b/c/d.

This rule set and request should end with a best match of RW /a/b/c/d, Rule 2.

>    Example 3:
>    Rule 1: /a/b/c -> read_only deny
>    Rule 2: /a/b/c -> read_write allow
>    Is the user allow to read /a/b/c? How about write?

This points out the need to separately define R and W and not conjoin RW, though I think
we need to permit setting one, or both in the same policy, otherwise there's the need
to specify the same rule 2x, one with read and one with write, for some usage patterns.

> 3. What happen if a user matches both a user rule and a group rule?
>    Example 1:
>    Rule 1: /a/b/c -> user read_write deny
>    Rule 2: /a/b/c -> group read_only allow
>    If a user matches both rules, do they have read access on /a/b/c?

The user rule is the most specific, so that rule would apply.
In this case user has RW access to /a/b/c

>    Example 2:
>    Rule 1: /a/b/c -> user read_write deny
>    Rule 2: /a/b/c -> group read_write allow
>    If a user matches both rules, do they have read access on /a/b/c?
>    The intention of the rule is that the group has write access except one user.

Also, in this case, user is denied RW access to /a/b/c, because the user rule is the most specific.

> 4. We also have wildcard. What is "best match" mean?
>    Example 1:
>    Rule 1: /a/b[name=mgmt_intf]/c -> deny
>    Rule 2: /a/b[name=*]/c/d -> allow
>
>    If we have a request of /a/b[name=mgmt_intf]/c/d, it matches both rules.
>    Rule 2 has longer prefix. However, the intention of the rule is that we deny
>    access on the management interface (mgmt_intf).

I believe the user would not have access to the mgmt_intf/c content, but would be able to view the /d level data, yes.
and for all other interfaces c/d would be available.

>    Example 2:
>    Rule 1: /a/b[k1=*]/c/d[k2=v2]/e -> deny
>    Rule 2: /a/b[k1=v1]/c/d[k2=*]/e -> allow
>    If we have a request of /a/b[k1=v1]/c/d[k2=v2]/e, it matches both rules.

Specified keys are a longer match than a wildcard (this I'll add to the documentation).
So in this case the user would be denied access.

>    Example 3:
>    Rule 1: /a/b[k1=*][k2=v2]/c/d -> deny
>    Rule 2: /a/b[k1=v1][k2=*]/c/d -> allow
>    Multiple keys in the same path element is rare, but it is possible in the proto.
>    A request of /a/b[k1=v1][k2=v2]/c/d will match both rules.

As above I think this request would be denied.
I think these both end as denials because the most specific path is the deny
path in both cases.

Reply:
Seems like markdown remotes the "*" in my comment.
There are still some ambiguity. I suggest to have a section on detailed steps to apply the config in the other comment.

Reply:
oh :( it also changed my \* into \_ :( argh, markdown!

Reply:
note: backslash<star> works fine: \*

#sillymarkdownescapes